cloud/logstream.go

```diff
@@ -40,30 +43,28 @@ func (c *client) StreamLogs(ctx context.Context, buildID string, deltasCh chan [
 	})
 	eg.Go(func() error {
 		for {
-			select {
-			case <-ctx.Done():
-				return ctx.Err()
-			case deltas, ok := <-deltasCh:
-				if !ok {
-					err := streamClient.Send(&logstream.StreamLogRequest{
-						BuildId: buildID,
-						Eof:     true,
-					})
-					if err != nil {
-						return errors.Wrap(err, "failed to send EOF to log stream")
-					}
-					mu.Lock()
-					finished = true
-					mu.Unlock()
-					return nil
-				}
-				err := streamClient.Send(&logstream.StreamLogRequest{
+			dl, err := deltas.Next(ctx)
+			if errors.Is(err, io.EOF) {
+				msg := &logstream.StreamLogRequest{
 					BuildId: buildID,
-					Deltas:  deltas,
-				})
-				if err != nil {
-					return errors.Wrap(err, "failed to send log delta")
+					Eof:     true,
+				}
+				if err := streamClient.Send(msg); err != nil {
```

Comment:
Nit: I generally prefer not putting things that have side effects in `if` expressions. It helps with readability. There are pros and cons regarding the scope of `err` created here. Not a very strong opinion.

Reply:
Yeah, my counter-point is specifically about the `err` variable and go's dumb typed nils. sticking to `if err := ...; err != nil` helps avoid the typed vs untyped nil problems. But I don't have strong opinions - as long as I can still use `:=` instead of reassignment.